File: ios/Outbound/Outbound/Domains/Athlete/CompanionContracts.swift
import Foundation

enum CompanionTask: String, Codable, Sendable {
    case answerTrainingQuestion = "answer_training_question"
    case adaptToday = "adapt_today"
    case prepareWeek = "prepare_week"
    case postRunReflection = "post_run_reflection"
    case liveGuidance = "live_guidance"
    case inspectMemory = "inspect_memory"
    case productHelp = "product_help"
}

enum CompanionSurface: String, Codable, Sendable {
    case assistant
    case today
    case postRun = "post_run"
    case weeklyReview = "weekly_review"
    case liveSession = "live_session"
    case memory
}

struct CompanionPriorMessageDTO: Codable, Sendable {
    let role: String
    let text: String
}

struct CompanionSituationalSignalDTO: Codable, Sendable {
    let idempotencyKey: String
    let type: String
    let value: String
    let source: String
    let confidence: Double
    let privacy: String
    let consequenceLevel: String
    let possibleEffects: [String]
    let scope: [String: String]
    let observedAt: Date
    let freshUntil: Date
}

struct CompanionTurnRequestDTO: Codable, Sendable {
    let task: CompanionTask
    let surface: CompanionSurface
    let prompt: String
    let conversationKey: String
    let recentMessages: [CompanionPriorMessageDTO]
    let currentEntityIds: [String]
    let clientCapabilities: [String]
    let isOffline: Bool
    let timeZoneIdentifier: String?
    let signals: [CompanionSituationalSignalDTO]
}

struct CompanionActionDTO: Codable, Identifiable, Sendable {
    let id: String
    let actionType: String
    let permissionTier: Int
    let requiresConfirmation: Bool
    let status: String
    let explanation: String
}

struct CompanionConfirmationDTO: Codable, Sendable {
    let actionId: String
    let title: String
    let explanation: String
    let acceptLabel: String
    let rejectLabel: String
}

struct CompanionContextReceiptDTO: Codable, Sendable {
    let manifestId: String
    let task: CompanionTask
    let tokenBudget: Int
    let estimatedTokens: Int
    let includedReferenceCount: Int
}

struct CompanionTurnResponseDTO: Codable, Sendable {
    let message: String
    let locale: String?
    let action: CompanionActionDTO?
    let confirmationRequest: CompanionConfirmationDTO?
    let suggestedReplies: [String]
    let runnerModelVersion: String
    let contextReceipt: CompanionContextReceiptDTO
}

struct CompanionActionDecisionRequestDTO: Codable, Sendable {
    let decision: String
}

struct CompanionActionDecisionResponseDTO: Codable, Sendable {
    let action: CompanionDecidedActionDTO
}

struct CompanionDecidedActionDTO: Codable, Sendable {
    let id: String
    let status: String
    let explanation: String
}

struct CompanionMemoryDTO: Codable, Identifiable, Sendable {
    var id: String { stableKey }
    let stableKey: String
    let kind: String
    let label: String
    let summary: String
    let confidence: Double
    let status: String
    let source: String
    let evidenceCount: Int
    let contradictionCount: Int
    let sensitivity: String
    let consequenceLevel: String
    let refreshedAt: Date
    let expiresAt: Date?
}

struct CompanionMemoriesResponseDTO: Codable, Sendable {
    let memories: [CompanionMemoryDTO]
}

struct CompanionMemoryCorrectionRequestDTO: Codable, Sendable {
    let value: String
    let summary: String
    let label: String?
    let idempotencyKey: String
}

struct CompanionMemoryCorrectionResponseDTO: Codable, Sendable {
    let memory: CompanionMemoryDTO
}

struct CompanionMemoryForgetRequestDTO: Codable, Sendable {
    let idempotencyKey: String
}

struct CompanionMemoryForgetResponseDTO: Codable, Sendable {
    let forgotten: Bool
}

struct CompanionSessionBriefDTO: Codable, Sendable {
    let version: Int
    let runnerModelVersion: String
    let workout: CompanionSessionWorkoutDTO?
    let readiness: CompanionSessionReadinessDTO?
    let guidancePriorities: [String]
    let cuePreferences: [String]
    let forbiddenBehavior: [String]
}

struct CompanionSessionWorkoutDTO: Codable, Sendable {
    let id: String
    let title: String
    let purpose: String
    let durationSeconds: Int
}

struct CompanionSessionReadinessDTO: Codable, Sendable {
    let choice: String?
    let energy: Int
    let soreness: Int
    let illnessOrPain: Bool
}

File: android/core/assistant/src/main/kotlin/com/plainstride/outbound/core/assistant/CompanionContracts.kt
package com.plainstride.outbound.core.assistant

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable enum class CompanionTask {
    @SerialName("answer_training_question") AnswerTrainingQuestion,
    @SerialName("adapt_today") AdaptToday,
    @SerialName("prepare_week") PrepareWeek,
    @SerialName("post_run_reflection") PostRunReflection,
    @SerialName("live_guidance") LiveGuidance,
    @SerialName("inspect_memory") InspectMemory,
    @SerialName("product_help") ProductHelp,
}

@Serializable enum class CompanionSurface {
    @SerialName("assistant") Assistant,
    @SerialName("today") Today,
    @SerialName("post_run") PostRun,
    @SerialName("weekly_review") WeeklyReview,
    @SerialName("live_session") LiveSession,
    @SerialName("memory") Memory,
}

@Serializable data class CompanionMessage(val role: String, val text: String, val createdAt: String? = null)
@Serializable data class CompanionTurnRequest(
    val task: CompanionTask = CompanionTask.AnswerTrainingQuestion,
    val surface: CompanionSurface = CompanionSurface.Assistant,
    val prompt: String,
    val conversationKey: String = "android-assistant",
    val recentMessages: List<CompanionMessage> = emptyList(),
    val currentEntityIds: List<String> = emptyList(),
    val clientCapabilities: List<String> = listOf("typed-action", "confirmation"),
    val isOffline: Boolean = false,
    val timeZoneIdentifier: String? = null,
    val signals: List<CompanionSignal> = emptyList(),
)
@Serializable data class CompanionSignal(val idempotencyKey: String, val type: String, val value: kotlinx.serialization.json.JsonElement, val source: String, val confidence: Double = 1.0, val privacy: String = "standard", val consequenceLevel: String = "low", val possibleEffects: List<String> = emptyList(), val scope: Map<String, String> = emptyMap(), val observedAt: String, val freshUntil: String)
@Serializable data class CompanionAction(val id: String, val actionType: String, val permissionTier: Int, val requiresConfirmation: Boolean, val status: String, val explanation: String)
@Serializable data class CompanionConfirmation(val actionId: String, val title: String, val explanation: String, val acceptLabel: String, val rejectLabel: String)
@Serializable data class CompanionContextReceipt(val manifestId: String, val task: CompanionTask, val tokenBudget: Int, val estimatedTokens: Int, val includedReferenceCount: Int)
@Serializable data class CompanionTurnResponse(val message: String, val locale: String? = null, val action: CompanionAction? = null, val confirmationRequest: CompanionConfirmation? = null, val suggestedReplies: List<String> = emptyList(), val runnerModelVersion: String, val contextReceipt: CompanionContextReceipt)
@Serializable data class CompanionDecisionRequest(val decision: String)
@Serializable data class CompanionDecidedAction(val id: String, val status: String, val explanation: String)
@Serializable data class CompanionDecisionResponse(val action: CompanionDecidedAction)

data class CompanionConversationState(
    val messages: List<CompanionMessage> = emptyList(),
    val sending: Boolean = false,
    val confirmation: CompanionConfirmation? = null,
    val suggestedReplies: List<String> = emptyList(),
    val lastFailure: String? = null,
)
